src/beast/beast/asio/async_completion.h

```diff
@@ -0,0 +1,52 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of Beast: https://github.com/vinniefalco/Beast
+    Copyright 2013, Vinnie Falco <vinnie.falco@gmail.com>
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef BEAST_ASIO_ASYNC_COMPLETION_H_INLUDED
+#define BEAST_ASIO_ASYNC_COMPLETION_H_INLUDED
+
+#include <beast/asio/type_check.h>
+#include <boost/asio/async_result.hpp>
+#include <boost/asio/handler_type.hpp>
+#include <type_traits>
+#include <utility>
+
+namespace beast {
+
+template <class CompletionToken, class Signature>
+struct async_completion
+{
+    using handler_type =
+        typename boost::asio::handler_type<
+            CompletionToken, Signature>::type;
+
+    async_completion(std::remove_reference_t<CompletionToken>& token)
+        : handler(std::forward<CompletionToken>(token))
```

Comment:
Scott is correct that `std::forward` doesn't do anything.  `token` is always forwarded as an lvalue here.  The N3896 docs use `forward` correctly in that `Handler` is a forwarding reference: `Handler&& handler`.  But `CompletionToken` is not a forwarding reference.  `token` is declared as an ordinary lvalue reference.

As written, I think this code can be simplified to this with no change in functionality:
````
: handler(token)
````
The difference is that the N3896 will move from `handler` if an rvalue handler is supplied by the client.  With `async_completion`, if an rvalue `token` is supplied by the client, we will get a compile-time error (except on VS probably).